contracts/zero-ex/contracts/src/features/LimitOrdersFeature.sol

```diff
@@ -0,0 +1,875 @@
+    ///      attached to this call. Any unspent ETH will be refunded to
+    ///      `msg.sender` (not `sender`).
+    /// @param order The limit order.
+    /// @param signature The order signature.
+    /// @param takerTokenFillAmount Maximum taker token to fill this order with.
+    /// @param taker The order taker.
+    /// @param sender The order sender.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _fillLimitOrder(
+        LibLimitOrder.LimitOrder memory order,
+        LibSignature.Signature memory signature,
+        uint128 takerTokenFillAmount,
+        address taker,
+        address sender
+    )
+        public
+        override
+        payable
+        onlySelf
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        uint256 ethProtocolFeePaid;
+        (ethProtocolFeePaid, takerTokenFilledAmount, makerTokenFilledAmount) =
+            _fillLimitOrderPrivate(FillLimitOrderPrivateParams({
+                order: order,
+                signature: signature,
+                takerTokenFillAmount: takerTokenFillAmount,
+                taker: taker,
+                sender: sender
+            }));
+        _refundProtocolFeeToSender(ethProtocolFeePaid);
+    }
+
+    /// @dev Fill an RFQ order. Internal variant. ETH protocol fees can be
+    ///      attached to this call. Any unspent ETH will be refunded to
+    ///      `msg.sender` (not `sender`).
+    /// @param order The RFQ order.
+    /// @param signature The order signature.
+    /// @param takerTokenFillAmount Maximum taker token to fill this order with.
+    /// @param taker The order taker.
+    /// @return takerTokenFilledAmount How much maker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _fillRfqOrder(
+        LibLimitOrder.RfqOrder memory order,
+        LibSignature.Signature memory signature,
+        uint128 takerTokenFillAmount,
+        address taker
+    )
+        public
+        override
+        payable
+        onlySelf
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        uint256 ethProtocolFeePaid;
+        (ethProtocolFeePaid, takerTokenFilledAmount, makerTokenFilledAmount) =
+            _fillRfqOrderPrivate(
+                order,
+                signature,
+                takerTokenFillAmount,
+                taker
+            );
+        _refundProtocolFeeToSender(ethProtocolFeePaid);
+    }
+
+
+    /// @dev Cancel a single limit order. The caller must be the maker.
+    ///      Silently succeeds if the order has already been cancelled.
+    /// @param order The limit order.
+    function cancelLimitOrder(LibLimitOrder.LimitOrder memory order)
+        public
+        override
+    {
+        bytes32 orderHash = getLimitOrderHash(order);
```

Comment:
I was just thinking... What if we indexed the `filled` mapping by the maker 🤔. Then we could just call `cancelLimitOrder` with the order hash and probably save a ton of gas.. So long as it would not make fills much more expensive -- I guess one additional `keccak` to lookup the cancelled mapping.